File: python/ltx/src/interplanet_ltx/_merkle.py
"""
_merkle.py — RFC 9162-style Merkle Audit Log
Story 28.5 — LTX Python SDK Merkle tree implementation

Leaf hash:  SHA-256(0x00 || entry_bytes)
Node hash:  SHA-256(0x01 || left || right)
Empty root: 32 zero bytes (64 hex zeros)
"""

from __future__ import annotations

import base64
import hashlib
from datetime import datetime, timezone
from typing import Any, Dict, List

from ._security import canonical_json

# ── Constants ─────────────────────────────────────────────────────────────────

EMPTY_ROOT = bytes(32)


# ── Internal hash helpers ─────────────────────────────────────────────────────


def _leaf_hash(entry_bytes: bytes) -> bytes:
    """SHA-256(0x00 || entry_bytes) — prevents second-preimage attacks."""
    return hashlib.sha256(b'\x00' + entry_bytes).digest()


def _node_hash(left: bytes, right: bytes) -> bytes:
    """SHA-256(0x01 || left || right) — prevents second-preimage attacks."""
    return hashlib.sha256(b'\x01' + left + right).digest()


def _largest_power_of_2_less_than(n: int) -> int:
    """Return the largest power of 2 strictly less than n (n >= 2)."""
    # (n-1).bit_length() - 1 gives floor(log2(n-1)), so 2**(that) is the result.
    return 1 << ((n - 1).bit_length() - 1)


def _root(leaves: List[bytes]) -> bytes:
    """Recursively compute the Merkle root of a slice of leaf hashes."""
    if not leaves:
        return EMPTY_ROOT
    if len(leaves) == 1:
        return leaves[0]
    mid = _largest_power_of_2_less_than(len(leaves))
    return _node_hash(_root(leaves[:mid]), _root(leaves[mid:]))


# ── MerkleLog ────────────────────────────────────────────────────────────────


class MerkleLog:
    """
    RFC 9162-compatible Merkle audit log.

    Maintains an append-only list of leaf hashes and supports
    inclusion proofs, consistency proofs, and signed tree heads.
    """

    def __init__(self) -> None:
        self._leaves: List[bytes] = []

    def append(self, entry: Any) -> Dict[str, Any]:
        """
        Append an entry to the log.

        Parameters
        ----------
        entry : Any
            Any JSON-serialisable object.

        Returns
        -------
        dict
            ``{'tree_size': int, 'root': str}``
        """
        entry_bytes = canonical_json(entry).encode('utf-8')
        self._leaves.append(_leaf_hash(entry_bytes))
        return {'tree_size': len(self._leaves), 'root': _root(self._leaves).hex()}

    def tree_size(self) -> int:
        """Return the current number of log entries."""
        return len(self._leaves)

    def root_hex(self) -> str:
        """Return the current Merkle root as a 64-character hex string."""
        return _root(self._leaves).hex()

    def inclusion_proof(self, leaf_index: int) -> List[Dict[str, str]]:
        """
        Compute an inclusion proof for the leaf at ``leaf_index`` (0-based).

        Parameters
        ----------
        leaf_index : int
            Zero-based index of the leaf.

        Returns
        -------
        list of dict
            Each element has ``{'side': 'left'|'right', 'hash': str}``.

        Raises
        ------
        IndexError
            If ``leaf_index`` is out of range.
        """
        if leaf_index >= len(self._leaves):
            raise IndexError('leaf index out of range')
        proof: List[Dict[str, str]] = []

        def build(lo: int, hi: int, idx: int) -> None:
            if hi - lo == 1:
                return
            mid = _largest_power_of_2_less_than(hi - lo)
            if idx < lo + mid:
                build(lo, lo + mid, idx)
                proof.append({'side': 'right', 'hash': _root(self._leaves[lo + mid:hi]).hex()})
            else:
                build(lo + mid, hi, idx)
                proof.append({'side': 'left', 'hash': _root(self._leaves[lo:lo + mid]).hex()})

        build(0, len(self._leaves), leaf_index)
        return proof

    def verify_inclusion(
        self,
        entry: Any,
        leaf_index: int,
        proof: List[Dict[str, str]],
        known_root: str,
    ) -> bool:
        """
        Verify an inclusion proof for ``entry`` against ``known_root``.

        Parameters
        ----------
        entry : Any
            The original entry (must produce the same canonical JSON).
        leaf_index : int
            Zero-based index (unused in computation, kept for API symmetry).
        proof : list of dict
            Output from :meth:`inclusion_proof`.
        known_root : str
            The trusted Merkle root (hex string).

        Returns
        -------
        bool
            ``True`` if the proof is valid.
        """
        h = _leaf_hash(canonical_json(entry).encode())
        for step in proof:
            sibling = bytes.fromhex(step['hash'])
            if step['side'] == 'right':
                h = _node_hash(h, sibling)
            else:
                h = _node_hash(sibling, h)
        return h.hex() == known_root

    def consistency_proof(self, old_size: int) -> List[str]:
        """
        Compute a consistency proof showing the current tree is an
        extension of a tree that had ``old_size`` entries.

        Parameters
        ----------
        old_size : int
            The size of the smaller (older) tree.

        Returns
        -------
        list of str
            List of hex-encoded hash strings forming the proof.

        Raises
        ------
        ValueError
            If ``old_size`` exceeds the current tree size.
        """
        new_size = len(self._leaves)
        if old_size > new_size:
            raise ValueError('old_size > new_size')
        if old_size == new_size:
            return []
        proof: List[str] = []

        def build(lo: int, hi: int, old_hi: int, first: bool) -> None:
            if lo == hi:
                return
            if lo + 1 == hi:
                if not first:
                    proof.append(self._leaves[lo].hex())
                return
            mid = _largest_power_of_2_less_than(hi - lo)
            if old_hi - lo <= mid:
                proof.append(_root(self._leaves[lo + mid:hi]).hex())
                build(lo, lo + mid, old_hi, first)
            else:
                if not first:
                    proof.append(_root(self._leaves[lo:lo + mid]).hex())
                build(lo + mid, hi, old_hi, False)

        build(0, new_size, old_size, True)
        return proof

    def sign_tree_head(self, private_key_b64: str, node_id: str) -> Dict[str, Any]:
        """
        Sign the current tree head with an Ed25519 private key.

        Parameters
        ----------
        private_key_b64 : str
            Base64url-encoded raw 32-byte Ed25519 private seed (no padding).
        node_id : str
            The signer's node ID (stored in the signed head).

        Returns
        -------
        dict
            Signed tree head with keys:
            ``sha256RootHash``, ``signerNodeId``, ``timestamp``,
            ``treeSize``, ``treeHeadSig``.

        Raises
        ------
        RuntimeError
            If the ``cryptography`` library is not installed.
        """
        try:
            from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
        except ImportError:
            raise RuntimeError(
                'cryptography library required for signing; install: pip install cryptography'
            )

        head: Dict[str, Any] = {
            'sha256RootHash': self.root_hex(),
            'signerNodeId': node_id,
            'timestamp': datetime.now(timezone.utc).isoformat(),
            'treeSize': len(self._leaves),
        }
        head_str = canonical_json(head)

        raw_seed = base64.urlsafe_b64decode(private_key_b64 + '==')
        priv_key = Ed25519PrivateKey.from_private_bytes(raw_seed)
        sig = priv_key.sign(head_str.encode('utf-8'))
        sig_b64 = base64.urlsafe_b64encode(sig).rstrip(b'=').decode()

        return {**head, 'treeHeadSig': sig_b64}


# ── verify_tree_head ──────────────────────────────────────────────────────────


def verify_tree_head(signed_head: Dict[str, Any], nik: Dict[str, Any]) -> bool:
    """
    Verify a signed tree head produced by :meth:`MerkleLog.sign_tree_head`.

    Parameters
    ----------
    signed_head : dict
        Output from :meth:`MerkleLog.sign_tree_head`.
    nik : dict
        NIK record of the signer (must have ``publicKey``).

    Returns
    -------
    bool
        ``True`` if the signature is valid, ``False`` otherwise.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
    except ImportError:
        return False

    sig_b64 = signed_head.get('treeHeadSig', '')
    head = {k: v for k, v in signed_head.items() if k != 'treeHeadSig'}
    head_str = canonical_json(head)

    raw_pub = base64.urlsafe_b64decode(nik['publicKey'] + '==')
    pub_key = Ed25519PublicKey.from_public_bytes(raw_pub)

    sig_bytes = base64.urlsafe_b64decode(sig_b64 + '==')
    try:
        pub_key.verify(sig_bytes, head_str.encode('utf-8'))
        return True
    except InvalidSignature:
        return False

File: typescript/ltx/src/merkle.ts
/**
 * @interplanet/ltx — Merkle Audit Log
 * Story 28.5 — RFC 9162-style Merkle tree audit log
 *
 * Leaf hash:  SHA-256(0x00 || entry_bytes)
 * Node hash:  SHA-256(0x01 || left || right)
 * Empty root: 32 zero bytes (64 hex zeros)
 */

import * as nodeCrypto from 'node:crypto';
import { canonicalJSON } from './security.js';
import type { NIK } from './security.js';

// ── Public interfaces ─────────────────────────────────────────────────────

export interface MerkleAppendResult {
  treeSize: number;
  root: string;
}

export interface InclusionProofStep {
  side: 'left' | 'right';
  hash: string;
}

export interface SignedTreeHead {
  treeSize: number;
  sha256RootHash: string;
  timestamp: string;
  signerNodeId: string;
  treeHeadSig: string;
}

export interface MerkleLog {
  append(entry: unknown): MerkleAppendResult;
  treeSize(): number;
  rootHex(): string;
  inclusionProof(leafIndex: number): InclusionProofStep[];
  verifyInclusion(
    entry: unknown,
    leafIndex: number,
    proof: InclusionProofStep[],
    knownRoot: string,
  ): boolean;
  consistencyProof(oldSize: number): string[];
  signTreeHead(privateKeyB64: string, nodeId: string): SignedTreeHead;
}

// ── Internal hash helpers ─────────────────────────────────────────────────

function _leafHash(entryBytes: Buffer): Buffer {
  const buf = Buffer.alloc(1 + entryBytes.length);
  buf[0] = 0x00;
  entryBytes.copy(buf, 1);
  return nodeCrypto.createHash('sha256').update(buf).digest() as Buffer;
}

function _nodeHash(left: Buffer, right: Buffer): Buffer {
  const buf = Buffer.alloc(1 + 32 + 32);
  buf[0] = 0x01;
  left.copy(buf, 1);
  right.copy(buf, 33);
  return nodeCrypto.createHash('sha256').update(buf).digest() as Buffer;
}

function _rootOf(leavesSlice: Buffer[]): Buffer {
  if (leavesSlice.length === 0) return Buffer.alloc(32);
  if (leavesSlice.length === 1) return leavesSlice[0];
  const mid = Math.pow(2, Math.floor(Math.log2(leavesSlice.length - 1)));
  return _nodeHash(
    _rootOf(leavesSlice.slice(0, mid)),
    _rootOf(leavesSlice.slice(mid)),
  );
}

// ── createMerkleLog ───────────────────────────────────────────────────────

/**
 * Create an RFC 9162-compatible Merkle audit log.
 *
 * @returns MerkleLog instance
 */
export function createMerkleLog(): MerkleLog {
  const leaves: Buffer[] = [];

  function root(): Buffer {
    return _rootOf(leaves.slice());
  }

  return {
    append(entry: unknown): MerkleAppendResult {
      const entryBytes = Buffer.from(canonicalJSON(entry), 'utf8');
      leaves.push(_leafHash(entryBytes));
      return { treeSize: leaves.length, root: root().toString('hex') };
    },

    treeSize(): number {
      return leaves.length;
    },

    rootHex(): string {
      return root().toString('hex');
    },

    inclusionProof(leafIndex: number): InclusionProofStep[] {
      if (leafIndex >= leaves.length) throw new Error('leaf index out of range');
      const proof: InclusionProofStep[] = [];

      function buildProof(lo: number, hi: number, idx: number): void {
        if (hi - lo === 1) return;
        const mid = Math.pow(2, Math.floor(Math.log2((hi - lo) - 1)));
        if (idx < lo + mid) {
          buildProof(lo, lo + mid, idx);
          proof.push({ side: 'right', hash: _rootOf(leaves.slice(lo + mid, hi)).toString('hex') });
        } else {
          buildProof(lo + mid, hi, idx);
          proof.push({ side: 'left', hash: _rootOf(leaves.slice(lo, lo + mid)).toString('hex') });
        }
      }

      buildProof(0, leaves.length, leafIndex);
      return proof;
    },

    verifyInclusion(
      entry: unknown,
      _leafIndex: number,
      proof: InclusionProofStep[],
      knownRoot: string,
    ): boolean {
      let hash = _leafHash(Buffer.from(canonicalJSON(entry), 'utf8'));
      for (const step of proof) {
        const sibling = Buffer.from(step.hash, 'hex');
        hash = step.side === 'right'
          ? _nodeHash(hash, sibling)
          : _nodeHash(sibling, hash);
      }
      return hash.toString('hex') === knownRoot;
    },

    consistencyProof(oldSize: number): string[] {
      const newSize = leaves.length;
      if (oldSize > newSize) throw new Error('oldSize > newSize');
      if (oldSize === newSize) return [];
      const proof: string[] = [];

      function buildConsistency(lo: number, hi: number, oldHi: number, first: boolean): void {
        if (lo === hi) return;
        if (lo + 1 === hi) {
          if (!first) proof.push(leaves[lo].toString('hex'));
          return;
        }
        const mid = Math.pow(2, Math.floor(Math.log2((hi - lo) - 1)));
        if (oldHi - lo <= mid) {
          proof.push(_rootOf(leaves.slice(lo + mid, hi)).toString('hex'));
          buildConsistency(lo, lo + mid, oldHi, first);
        } else {
          if (!first) proof.push(_rootOf(leaves.slice(lo, lo + mid)).toString('hex'));
          buildConsistency(lo + mid, hi, oldHi, false);
        }
      }

      buildConsistency(0, newSize, oldSize, true);
      return proof;
    },

    signTreeHead(privateKeyB64: string, nodeId: string): SignedTreeHead {
      const head = {
        sha256RootHash: root().toString('hex'),
        signerNodeId: nodeId,
        timestamp: new Date().toISOString(),
        treeSize: leaves.length,
      };
      const headStr = canonicalJSON(head);

      // Reconstruct Ed25519 private key from raw 32-byte seed via PKCS8 DER wrapping
      const rawSeed = Buffer.from(privateKeyB64, 'base64url');
      const pkcs8Header = Buffer.from('302e020100300506032b657004220420', 'hex');
      const pkcs8Der = Buffer.concat([pkcs8Header, rawSeed]);
      const privKey = nodeCrypto.createPrivateKey({ key: pkcs8Der, format: 'der', type: 'pkcs8' });

      const sigBytes = nodeCrypto.sign(null, Buffer.from(headStr, 'utf8'), privKey);

      return { ...head, treeHeadSig: sigBytes.toString('base64url') };
    },
  };
}

// ── verifyTreeHead ────────────────────────────────────────────────────────

/**
 * Verify a signed tree head produced by log.signTreeHead().
 *
 * @param signedHead  Output from signTreeHead()
 * @param nik         NIK record of the signer
 * @returns           true if the signature is valid
 */
export function verifyTreeHead(signedHead: SignedTreeHead, nik: NIK): boolean {
  const { treeHeadSig, ...head } = signedHead;
  const headStr = canonicalJSON(head);

  // Reconstruct Ed25519 public key from raw 32 bytes via SPKI DER wrapping
  const rawPub = Buffer.from(nik.publicKey, 'base64url');
  const spkiHeader = Buffer.from('302a300506032b6570032100', 'hex');
  const spkiDer = Buffer.concat([spkiHeader, rawPub]);
  const pubKey = nodeCrypto.createPublicKey({ key: spkiDer, format: 'der', type: 'spki' });

  const sigBytes = Buffer.from(treeHeadSig, 'base64url');
  return nodeCrypto.verify(null, Buffer.from(headStr, 'utf8'), pubKey, sigBytes);
}
